### PR TITLE
Faster commits

### PR DESCRIPTION
**Pull request by niik**
🌵 **This depends on #1403** 🌵 

**TL;DR: This cuts the time it takes to make a commit with ~900 unstaged files from 1 minute to 4 seconds**

As I was working on #1403 I tested committing using a repository with ~900 unstaged files. I created the repository by throwing in some node modules

```shell
git init lotsa-files
cd lotsa-files
npm init
npm install --save cheerio request got
```

That gave me 917 unstaged files to work with. Committing all of these in one swoop in Desktop took about 1 minute and 2 seconds on average while `git add .` took ~4s.

The reason for it taking so long was that we call out to `git add` for each added/modified/removed file. On Windows this incurs a _significant_ cost so I switched over to using `update-index` instead which lets you pass a list of files over stdin, allowing us to update the index in bulk. This cut the time down to 4.6 seconds on average, shaving almost a full minute from our commit time.

Now, a change like this isn't uncontroversial and I believe it should fall under the same question that was posed in https://github.com/desktop/desktop/pull/1397#issuecomment-299890017. One thing that speaks in favor of this PR landing (aside from the performance boost) is that the behavior we currently have implemented using add is unintuitive at best.

Please bear with me as I run through a fairly contrived example of a renamed file.

```shell
echo "foo" > foo && git add foo && git commit -m 'foo'
git mv foo bar
echo "bar" > foo
```

After this both apps show the following (I've only selected the rename change)

image

But after committing this change the results differ

image

The old version (using `git add`) commits the change file even though it wasn't selected, whereas the new version (using `git update-index`) correctly commits only the rename and leaves the modified source file untouched in the working directory.

image

### CODE REVIEW

**Comment by niik** [comment]
@joshaber Same question here as in #1435, when should we consider landing this?

File: app/src/lib/git/update-index.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+ *                status from the working directory.
+ * 
+ * @param options See the IUpdateIndexOptions interface for more details. 
+ */
+export async function updateIndex(repository: Repository, paths: ReadonlyArray<string>, options: IUpdateIndexOptions = { }) {
+
+  if (!paths.length) {
+    return
+  }
+  console.log(paths)
```
Comment (joshaber) [inline app/src/lib/git/update-index.ts]: :fire:

File: app/src/lib/git/update-index.ts [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+/**
+ * Updates the index with file contents from the working tree.
+ * 
+ * @param paths   A list of paths which are to be updated with file contents and
+ *                status from the working directory.
+ * 
+ * @param options See the IUpdateIndexOptions interface for more details. 
+ */
+export async function updateIndex(repository: Repository, paths: ReadonlyArray<string>, options: IUpdateIndexOptions = { }) {
```
Comment (joshaber) [inline app/src/lib/git/update-index.ts]: It doesn't look like we need to export this function.

**Comment by niik** [comment]
🐨
